File: server/src/main/java/com/krillsson/sysapi/core/metrics/defaultimpl/DefaultNetworkMetrics.java
/*
 * Sys-Api (https://github.com/Krillsson/sys-api)
 *
 * Copyright 2017 Christian Jensen / Krillsson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Maintainers:
 * contact[at]christian-jensen[dot]se
 */
package com.krillsson.sysapi.core.metrics.defaultimpl;

import com.krillsson.sysapi.core.metrics.NetworkMetrics;
import com.krillsson.sysapi.core.SpeedMeasurementManager;
import com.krillsson.sysapi.core.domain.network.NetworkInterface;
import com.krillsson.sysapi.core.domain.network.NetworkInterfaceLoad;
import com.krillsson.sysapi.core.domain.network.NetworkInterfaceMetrics;
import com.krillsson.sysapi.core.domain.network.NetworkInterfaceSpeed;
import org.slf4j.Logger;
import oshi.hardware.HardwareAbstractionLayer;
import oshi.hardware.NetworkIF;

import java.net.SocketException;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

public class DefaultNetworkMetrics implements NetworkMetrics {

    private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger(DefaultNetworkMetrics.class);

    private static final int BYTE_TO_BIT = 8;
    protected static final NetworkInterfaceSpeed EMPTY_INTERFACE_SPEED = new NetworkInterfaceSpeed(0, 0);

    private final HardwareAbstractionLayer hal;
    private final SpeedMeasurementManager speedMeasurementManager;

    protected DefaultNetworkMetrics(HardwareAbstractionLayer hal, SpeedMeasurementManager speedMeasurementManager) {
        this.hal = hal;
        this.speedMeasurementManager = speedMeasurementManager;
    }

    void register() {
        List<SpeedMeasurementManager.SpeedSource> collect = Arrays.stream(hal.getNetworkIFs())
                .map(n -> new SpeedMeasurementManager.SpeedSource() {
                    @Override
                    public String getName() {
                        return n.getName();
                    }

                    @Override
                    public long getCurrentRead() {
                        n.updateNetworkStats();
                        return n.getBytesRecv() * BYTE_TO_BIT;
                    }

                    @Override
                    public long getCurrentWrite() {
                        //TODO: maybe it's good enough to do this in getCurrentRead since getCurrentWrite is called immediately after
                        n.updateNetworkStats();
                        return n.getBytesSent() * BYTE_TO_BIT;
                    }
                })
                .collect(Collectors.toList());
        speedMeasurementManager.register(collect);
    }

    @Override
    public List<NetworkInterface> networkInterfaces() {
        return Arrays.stream(hal.getNetworkIFs()).map(mapToNetworkInterface()).collect(Collectors.toList());
    }

    @Override
    public Optional<NetworkInterface> networkInterfaceById(String id) {
        return Arrays.stream(hal.getNetworkIFs()).filter(n -> n.getName().equalsIgnoreCase(id)).map(
                mapToNetworkInterface()).findAny();
    }

    @Override
    public List<NetworkInterfaceLoad> networkInterfaceLoads() {
        return Arrays.stream(hal.getNetworkIFs())
                .map(mapToLoad())
                .collect(Collectors.toList());
    }

    @Override
    public Optional<NetworkInterfaceLoad> networkInterfaceLoadById(String id) {
        return Arrays.stream(hal.getNetworkIFs())
                .filter(n -> n.getName().equalsIgnoreCase(id))
                .map(mapToLoad())
                .findAny();
    }


    protected NetworkInterfaceSpeed speedForInterfaceWithName(String name) {
        Optional<SpeedMeasurementManager.CurrentSpeed> currentSpeedForName = speedMeasurementManager.getCurrentSpeedForName(
                name);
        return currentSpeedForName.map(s -> new NetworkInterfaceSpeed(
                s.getReadPerSeconds(),
                s.getWritePerSeconds()
        )).orElse(EMPTY_INTERFACE_SPEED);
    }


    Function<NetworkIF, NetworkInterface> mapToNetworkInterface() {
        return nic -> {
            boolean loopback = false;
            try {
                loopback = nic.getNetworkInterface().isLoopback();
            } catch (SocketException e) {
                //ignore
                LOGGER.warn("Socket exception while queering for loopback parameter", e);
            }
            return new NetworkInterface(
                    nic.getName(),
                    nic.getDisplayName(),
                    nic.getMacaddr(),
                    nic.getMTU(),
                    loopback,
                    nic.getIPv4addr(),
                    nic.getIPv6addr()
            );
        };
    }

    Function<NetworkIF, NetworkInterfaceLoad> mapToLoad() {
        return n -> new NetworkInterfaceLoad(
                new NetworkInterfaceMetrics(
                        n.getSpeed(),
                        n.getBytesRecv(),
                        n.getBytesSent(),
                        n.getPacketsRecv(),
                        n.getPacketsSent(),
                        n.getInErrors(),
                        n.getOutErrors()
                ),
                speedForInterfaceWithName(n.getName())
        );
    }
}

File: server/src/main/java/com/krillsson/sysapi/core/metrics/defaultimpl/DefaultNetworkMetrics.kt
/*
 * Sys-Api (https://github.com/Krillsson/sys-api)
 *
 * Copyright 2017 Christian Jensen / Krillsson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Maintainers:
 * contact[at]christian-jensen[dot]se
 */
package com.krillsson.sysapi.core.metrics.defaultimpl

import com.krillsson.sysapi.core.domain.network.NetworkInterface
import com.krillsson.sysapi.core.domain.network.NetworkInterfaceLoad
import com.krillsson.sysapi.core.domain.network.NetworkInterfaceSpeed
import com.krillsson.sysapi.core.domain.network.NetworkInterfaceValues
import com.krillsson.sysapi.core.metrics.NetworkMetrics
import com.krillsson.sysapi.core.speed.SpeedMeasurementManager
import com.krillsson.sysapi.core.speed.SpeedMeasurementManager.CurrentSpeed
import com.krillsson.sysapi.core.speed.SpeedMeasurementManager.SpeedSource
import org.slf4j.LoggerFactory
import oshi.hardware.HardwareAbstractionLayer
import oshi.hardware.NetworkIF
import java.net.SocketException
import java.util.Optional

open class DefaultNetworkMetrics(
    private val hal: HardwareAbstractionLayer,
    private val speedMeasurementManager: SpeedMeasurementManager
) : NetworkMetrics {

    class NetworkSpeedSource(private val networkIF: NetworkIF) : SpeedSource {
        override fun getName(): String {
            return networkIF.name
        }

        override fun getCurrentRead(): Long {
            networkIF.updateAttributes()
            return networkIF.bytesRecv
        }

        override fun getCurrentWrite(): Long {
            networkIF.updateAttributes()
            return networkIF.bytesSent
        }
    }

    fun register() {
        speedMeasurementManager.register(
            hal.networkIFs.map {
                NetworkSpeedSource(it)
            }
        )
    }

    override fun networkInterfaces(): List<NetworkInterface> {
        return hal.networkIFs
            .map {
                var isLoopback = false
                try {
                    isLoopback = it.queryNetworkInterface().isLoopback
                } catch (e: SocketException) {
                    //ignore
                    LOGGER.warn("Socket exception while queering for loopback parameter", e)
                }
                it.asNetworkInterface(isLoopback)
            }.toList()
    }

    override fun networkInterfaceById(id: String): Optional<NetworkInterface> {
        return Optional.ofNullable(hal.networkIFs.firstOrNull {
            it.name.equals(
                id,
                ignoreCase = true
            )
        }
            ?.let {
                var isLoopback = false
                try {
                    isLoopback = it.queryNetworkInterface().isLoopback
                } catch (e: SocketException) {
                    //ignore
                    LOGGER.warn("Socket exception while queering for loopback parameter", e)
                }
                it.asNetworkInterface(isLoopback)
            })
    }

    override fun networkInterfaceLoads(): List<NetworkInterfaceLoad> {
        return hal.networkIFs.map {
            var up = false
            try {
                up = it.queryNetworkInterface().isUp
            } catch (e: SocketException) {
                LOGGER.error("Error occurred while getting status for NIC", e)
            }
            it.asNetworkInterfaceLoad(up, speedForInterfaceWithName(it.name))
        }.toList()
    }

    override fun networkInterfaceLoadById(id: String): Optional<NetworkInterfaceLoad> {
        return Optional.ofNullable(hal.networkIFs.firstOrNull {
            it.name.equals(
                id,
                ignoreCase = true
            )
        }
            ?.let {
                var up = false
                try {
                    up = it.queryNetworkInterface().isUp
                } catch (e: SocketException) {
                    LOGGER.error("Error occurred while getting status for NIC", e)
                }
                it.asNetworkInterfaceLoad(up, speedForInterfaceWithName(it.name))
            })
    }

    protected open fun speedForInterfaceWithName(name: String?): NetworkInterfaceSpeed {
        val currentSpeedForName = speedMeasurementManager.getCurrentSpeedForName(
            name
        )
        return currentSpeedForName.map { s: CurrentSpeed ->
            NetworkInterfaceSpeed(
                s.readPerSeconds,
                s.writePerSeconds
            )
        }.orElse(EMPTY_INTERFACE_SPEED)
    }

    private fun NetworkIF.asNetworkInterface(isLoopback: Boolean): NetworkInterface =
        NetworkInterface(
            name,
            displayName,
            macaddr,
            speed,
            mtu,
            isLoopback,
            iPv4addr.asList(),
            iPv6addr.asList()
        )

    private fun NetworkIF.asNetworkInterfaceLoad(
        up: Boolean,
        nicSpeed: NetworkInterfaceSpeed
    ): NetworkInterfaceLoad = NetworkInterfaceLoad(
        name,
        up,
        NetworkInterfaceValues(
            speed,
            bytesRecv,
            bytesSent,
            packetsRecv,
            packetsSent,
            inErrors,
            outErrors
        ),
        nicSpeed
    )

    companion object {
        @JvmField
        protected val EMPTY_INTERFACE_SPEED = NetworkInterfaceSpeed(0, 0)
        private val LOGGER = LoggerFactory.getLogger(
            DefaultNetworkMetrics::class.java
        )
        private const val BYTE_TO_BIT = 8
    }
}